typescript/end-to-end-testing/test/test.ts

```diff
@@ -0,0 +1,142 @@
+import * as net from "net";
+
+import { protoMetadata } from "../src/generated/proto/example";
+import { MyExampleService } from "../src/example_service";
+
+// Prepare the restate server
+async function prepareRestateServer(
+  mountServicesFn: (server: restate.RestateServer) => void
+): Promise<http2.Http2Server> {
+  // Prepare RestateServer
+  const restateServer = restate.createServer();
+  mountServicesFn(restateServer);
+
+  // Start HTTP2 server on random port
+  const restateHttpServer = http2.createServer(restateServer);
+  await new Promise((resolve, reject) => {
+    restateHttpServer
+      .listen(0)
+      .once("listening", resolve)
+      .once("error", reject);
+  });
+  const restateServerPort = (restateHttpServer.address() as net.AddressInfo)
+    .port;
+  console.info(`Started listening on port ${restateServerPort}`);
+
+  return restateHttpServer;
+}
+
+// Prepare the restate testcontainer
+async function prepareRestateTestContainer(
+  restateServerPort: number
+): Promise<StartedTestContainer> {
+  const restateContainer = new GenericContainer(
+    "docker.io/restatedev/restate:latest"
+  )
+    // Expose ports
+    .withExposedPorts(8080, 9070)
+    // Wait start on health checks
+    .withWaitStrategy(
+      Wait.forAll([
+        Wait.forHttp("/grpc.health.v1.Health/Check", 8080),
+        Wait.forHttp("/health", 9070),
+      ])
+    );
+
+  // This MUST be executed before starting the restate container
+  // Expose host port to access the restate server
+  await TestContainers.exposeHostPorts(restateServerPort);
+
+  // Start restate container
+  const startedRestateContainer = await restateContainer.start();
+
+  // From now on, if something fails, stop the container to cleanup the environment
+  try {
+    console.info("Going to register services");
+
+    // Register this service endpoint
+    const res = await fetch(
+      `http://${startedRestateContainer.getHost()}:${startedRestateContainer.getMappedPort(
+        9070
+      )}/deployments`,
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+          // See https://node.testcontainers.org/features/networking/#expose-host-ports-to-container
+          uri: `http://host.testcontainers.internal:${restateServerPort}`,
+        }),
+      }
+    );
+    if (!res.ok) {
+      const badResponse = await res.text();
+      throw new Error(
+        `Error ${res.status} during registration: ${badResponse}`
+      );
+    }
+
+    console.info("Registered");
+    return startedRestateContainer;
+  } catch (e) {
+    await startedRestateContainer.stop();
+    throw e;
+  }
+}
+
+describe("ExampleService", () => {
+  let startedRestateHttpServer: http2.Http2Server;
+  let startedRestateContainer: StartedTestContainer;
+  let clientTransport: Transport;
+
+  // Deploy Restate and the Service endpoint once for all the tests in this suite
+  beforeAll(async () => {
```

Comment:
Can we abstract this away so users can just copy over some test_setup.ts file, containing all this logic,
and then in their test just do something like: `beforeAll (async () => {clientTransport = bringUpSetup()})` ?

And `afterAll(async () => {await bringDownSetup()})` ?
This way we could easily document it as well.